### ISSUE TITLE
Agent: Garbage collect and filter unused requests and claims

### ISSUE THREAD

**Issue by fredo** [beamer-bridge/beamer#356]
We started with removing unfillable requests from the `RequestTracker`. We can improve on that by removing all information which are not of interest for the agent.
I would like to raise this as its own issue because we need to be careful about what information are not of interest for the agent. I.e. an expired request can still be relevant if there are challenged (or even pending) claims to it where the informaiton might be interesting for the agent (part of the challenge).
We can use this issue to come up with a good strategy.

### Todo
- [x] Think through that anything that is removed won't be necessary for the agent, e.g. never remove requests when there is an active claim.
- [x] Remove "all information which are not of interest for the agent".

**Comment by fredo** [comment]
My concerns in the refinement meeting are derived from the architecture @istankovic and me discussed last week back in the office.
We agreed that the state machines of claims and requests should be viewed as two separate state machines running in parallel completely detached from each other. Since claims are determined by other users, theoretically claims can arrive at any state of the request (but withdrawn). Although the request is already withdrawn, there could still be active claims, which need the requests internal information.

# Rules of the interaction of claims and requests
In order to keep control over the state machines we determined a few rules which we must follow in the future.

1) Claims have read access to requests' states and vice versa.
2) Inside of processing one state machine we should never write and/or trigger state changes of the other. We decided that to be able to retrace how a claim or request arrived in this state. Ideally in the future, state changes should only be triggered by events. Currently we accept the exception that we allow internal state changes when processing requests or states.


### Why we need to be cautious
Although, we have a final state for requests and claims (namely the withdrawn state), we do allow read access between the state machines (as described above). Thus, we cannot simply remove an object when it arrives in its final state, because the other state machine might need to read from this object still.

#### Example: Active claims
In theory, we can have an infinite number of active claims for a given request. While the request can be withdrawn by a successful claim OR L1 resolution, this does not withdraw all other active claims. In the agent's internal state this means, that there can exist active claims, which are not yet in the withdrawn state, but the request is already withdrawn and in its final state.
Those active claims might need to read the state of the finalized request still. We need to respect this when it comes to removing information from the internal state.

### What to do now?
We should take a close look to the claim and request state machine and check the times when there is read access between them. Then, formulate conditions when removal of information is safe. Implement these conditions when it comes to removal (final state).

**Comment by palango** [comment]
For claims I only find this reference on the request, which is why we need to wait with removal of a certain request until all related claims are removed.
https://github.com/beamer-bridge/beamer/blob/63e248f0f449f4bba0137f5f031a73e6e3eaab6f/beamer/chain.py#L234

I didn't find any access of claims from the requests state machine.
